lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -418,6 +453,55 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
     return success();
   }
 
+  LogicalResult handlePublicModuleRefPorts(FModuleOp module) {
+    // Decls must be at the top mlir module scope
+    auto declBuilder = ImplicitLocOpBuilder::atBlockBegin(
+        module.getLoc(), getOperation().getBody());
+    auto builder = ImplicitLocOpBuilder::atBlockBegin(module.getLoc(),
+                                                      circuitOp.getBodyBlock());
+
+    for (size_t portIndex = 0, numPorts = module.getNumPorts();
+         portIndex != numPorts; ++portIndex) {
+      auto refType = module.getPortType(portIndex).dyn_cast<RefType>();
+      if (!refType || isZeroWidth(refType.getType()) ||
+          module.getPortDirection(portIndex) != Direction::Out)
+        continue;
+      auto portValue =
+          module.getArgument(portIndex).cast<mlir::TypedValue<RefType>>();
+
+      Attribute ref;
+      SmallString<128> stringLeaf;
+      if (failed(resolveReferencePath(portValue, builder, ref, stringLeaf)))
+        return failure();
+
+      SmallString<128> formatString;
+      if (ref)
+        formatString += "{{0}}";
+      formatString += stringLeaf;
+
+      // Insert a macro with the format:
+      // ref_<circuit-name>_<module-name>_<ref-name> <path>
+      auto macroName = builder.getStringAttr("ref_" + circuitOp.getName() +
+                                             "_" + module.getName() + "_" +
+                                             module.getPortName(portIndex));
+      declBuilder.create<sv::MacroDeclOp>(macroName, ArrayAttr(), StringAttr());
+
+      auto macroDefOp = builder.create<sv::MacroDefOp>(
+          FlatSymbolRefAttr::get(macroName),
+          builder.getStringAttr(formatString),
+          builder.getArrayAttr(ref ? ref : ArrayRef<Attribute>{}));
+
+      // The macro will be exported to a file with the format:
+      // ref_<circuit-name>_<module-name>.sv
+      macroDefOp->setAttr(
+          "output_file", hw::OutputFileAttr::getFromFilename(
+                             &getContext(), "ref_" + circuitOp.getName() + "_" +
+                                                module.getName() + ".sv"));
```

Comment:
nit: could you avoid creating `"ref_" + circuitOp.getName() + "_" + module.getName()` twice here and in `macroName` creation?

Reply:
Updated